### DISCUSSION TITLE
training using the GPU

### DISCUSSION THREAD

**Discussion by thematheusk3** [voicepaw/so-vits-svc-fork#779] [General]
Hello everyone o/

I am having trouble training using the GPU.

Win 11
RTX 3060
Python 3.10.6
Tensorflow version 2.10.0
Cuda 11.2
Cudnn 8.1

The graphics card is detected using the following code:
print("Num GPUs available", len(tf.config.experimental.list_physical_devices('GPU')))
tf.test.gpu_device_name()

However, during training, the output shows 'GPU available: False, used: False'.

**Comment by jiaxinf** [comment]
Okay, I figured it out after hours of investigation. You need to first confirm if CUDA is available on your PC, run these commands:

```
python
import torch
torch.cuda.is_available()
```
if it returns false, then you have some problems. 
My attempt is to:
1. install Anaconda
2. start your Anacoda ternimal, not your regular PC terminal
3. create conda environment `conda create -n env_name python=3.10`, and activate it `conda activate env_name`
5. install Pytorch from there `conda install pytorch torchvision torchaudio pytorch-cuda=11.8 -c pytorch -c nvidia`
6. run the list of 3 commands at the top again, and ensure it returns true.(if you start training now, you may get 'no module name torch' error)
7. install this project with `pip install -U so-vits-svc-fork`
8. remember to always run your training or hubert commands from Anaconda terminal

Start your Task Manager, go to GPU and toggle 'CUDA' to see if there's any input there.

**Reply by thematheusk3** [reply]
After following the steps, it worked (GPU available: True (cuda), used: True). 

Is it possible to know the average time for training 200 audios on an RTX 3060?


Thank you very much for the support xD

**Reply by jiaxinf** [reply]
No problem, glad to help! It takes ~30s for me to complete 1 epoch and I'm on Tesla T4, so you should be expecting something faster. But I heard to achieve the optimal result you may need ~25k steps, and **steps needed for 1 epoch = sample size/batch_size**, the training time also depends on `epoch_number` & `batch_size` in your config file.

**Reply by Saaaul** [reply]
Faced such an error, tell me, please, how can I solve it? Appears after svc train -t
3080TI + CUDA

RuntimeError: CUDA error: unspecified launch failure
CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
For debugging consider passing CUDA_LAUNCH_BLOCKING=1.
Compile with TORCH_USE_CUDA_DSA to enable device-side assertions.

**Reply by Saaaul** [reply]
It's decided! It was necessary to set the standard settings of the video card. I disabled the configured profile in MSI Afterburner and everything worked!

**Reply by Saaaul** [reply]
But I still come across this warning: UserWarning: Type d Storage is deprecated
What does it mean and how to solve it?

**Reply by jiaxinf** [reply]
Warnings do not affect usage for users, you should be fine. If you do dislike them, bring this up to the developers of this project. Probably an old version of the library or class is used in the code.

**Comment by onikaii** [comment]
@jiaxinf 
Thanks, your instructions worked. This is my first time training a model,
![Sin título](https://github.com/voicepaw/so-vits-svc-fork/assets/94697463/f71931d0-8009-4026-9252-af282b3ee208)
I am training with more than 1000 audios,
I have a rxt 3060ti (8gb),   "epochs": 400, "batch_size": 8 (if I enter something greater than 8 I get an error message) 
each epoch does it takes  ~120s.

cpu percentage usage does not exceed 25%, Is there any way to speed up the epochs?
or is there a way/formula to calculate epochs and batch_size for optimal performance?

**Reply by jiaxinf** [reply]
do you want to increase the learning_rate or perhaps virtual memory in advanced system setting to see if that helps? My batch_size is 16 so it takes fewer steps in one epoch, also I only have 200+ samples, and my GPU memory is 24GB. 1000/8 = 125 steps for 1 single epoch in your case so you may want to reduce the sample size or rent a cloud GPU that supports larger batch_size like 32 or 40.

**Comment by LROER** [comment]
I don't know why, but my GPU is still not being detected. 
![изображение](https://github.com/voicepaw/so-vits-svc-fork/assets/105774942/04a59c48-0b59-45f5-aad7-c7af79c53519)
but cuda is avalible 
![изображение](https://github.com/voicepaw/so-vits-svc-fork/assets/105774942/f080197b-98ac-4927-9884-ddd67f0a67db)
Do you have any ideas?"

**Reply by jiaxinf** [reply]
Does your GPU support cuda? run `nvidia-smi`, what's your cuda version?, plus 
```
import torch
torch.zeros(1).cuda()
```
what's your output?
Basically you need to ensure 1). GPU supports cuda, 2). if it supports, pytorch version installed is compatible with your GPU

**Reply by LROER** [reply]
i solved it. I blunted and did not install the CUDA Toolkit. But now, after night of training. I got this 
![изображение](https://github.com/voicepaw/so-vits-svc-fork/assets/105774942/03aef53d-d060-4080-a0b2-6645018b4f26)
my cuda version is 12.1

**Reply by jiaxinf** [reply]
Try to switch your installed CUDA version to 11.8 and ensure the toolkit etc. is also in 11.8.

**Reply by evilcrusher2** [reply]
I've followed these steps and have hit this issue pasted below. I had to go get 11.8 toolkit of CUDA and installed it. I've made sure the environment has it at the top of the appropriate variable in windows system env variables. In the check you posted above I am still getting false, but it shows up as available when running the training. Now I can't get it to recognize the TPU, IPU, and HPU. I'm tying to remove everything I can of 12.2 and reinstall 11.8. I'm replacing nstudio and what not on the cuda tool kit and trying to find how I can downgrade the other drivers if that doesn't work. But I have no idea if that will work. I'm using a RTX 4070 that I just purchased to do this type of work.



INFO: GPU available: True (cuda), used: True
           INFO     [20:08:38] GPU available: True (cuda), used: True                                                                               rank_zero.py:53
INFO: TPU available: False, using: 0 TPU cores
           INFO     [20:08:38] TPU available: False, using: 0 TPU cores                                                                             rank_zero.py:53
INFO: IPU available: False, using: 0 IPUs
           INFO     [20:08:38] IPU available: False, using: 0 IPUs                                                                                  rank_zero.py:53
INFO: HPU available: False, using: 0 HPUs
           INFO     [20:08:38] HPU available: False, using: 0 HPUs

**Reply by evilcrusher2** [reply]
So I've gone backwards to geforce 3.26 and driver down to 531.61 back in April 2023. No dice. Same thing and it's not utlizing the whole gpu like it does on my laptop with the 3070ti. I got this card because it's better architecture should make it quicker, and has more vram which means I can have larger batches at a time to speed things up a little. But it appears that it's barely utilizing outside of CUDA and forcing the heavy lifting on my AMD quad core that wasn't meant for this. Heck the only reason I didn't upgrade the CPU was because it's good enough for doing the regular computing and the GPU would be doing most of the heavy lifting as I edit video, graphics, and music mainly.
![HQJHGtS](https://github.com/voicepaw/so-vits-svc-fork/assets/104800981/bede3aa4-6813-42e6-9e61-276681631200)

**Reply by jiaxinf** [reply]
Was the 3070ti using TPU etc. in training? I wasn't using any of those in my training. Also, what's your batch_size value?

**Reply by evilcrusher2** [reply]
So I was finally able to get to my laptop with the 3070ti and run a training session with it to check it out. And you are correct in them not being used in training, as they weren't in the training module I just ran. Only the GPU was identified while the script was running.  I ran 100 epoch of a batch size of 7 in about 6 minutes from hitting enter on the command until finish. This is the system results from the laptop GPU while the training was running. note that cuda isn't in the drop down selector. 

![XEzvG5j](https://github.com/voicepaw/so-vits-svc-fork/assets/104800981/fe064812-0527-45e7-9571-1fdf232e5a37)

As well, the batch size on the 4070 rig is 16 because I know the ram can handle that. So I don't get why the 3070ti is having 100% of the gpu utilized while using almost nothing of the CPU, and the 4070 is doing the opposite.

I did notice that the firmware versions were from a different earlier date, back in March. I wonder if that is making the difference. I will be getting ready to do a fresh install of Windows on the 4070 rig on a different m2 drive with 1tb of space instead of a smaller 250gb OS drive that came with the rig. The m2 drive is there, just needs to have data moved over to my 3rd or and external drive. I'll try to find earlier drivers and see if maybe that is also causing the issue. The only thing after that would be to update the OS to Win 11 in an attempt to get the two environments almost identical.

**Reply by evilcrusher2** [reply]
> Was the 3070ti using TPU etc. in training? I wasn't using any of those in my training. Also, what's your batch_size value?


And this is on a fresh install of windows with an install not using the batch file to auto install everything. Yes, from what I can gather the batch size is in a way determining how much gpu gets utilized. I can't tell if its smaller or bigger batches that would be better when within the ram parameters. It's almost like when it gets geared up to start processing, the next batch has to be done and it's having to start from scratch and slows down.  

![T4W8tPg](https://github.com/voicepaw/so-vits-svc-fork/assets/104800981/16678a9b-47fd-465f-9305-201c5e70aa59)

I'm at a loss other than the possibility of older drivers worked better and this card is too new with too new of drivers? I don't get why the 3070ti is utilizing itself more than the 4070. Or could it literally be the CPU bottlenecking it?

**Reply by evilcrusher2** [reply]
I have noticed that 11.8 shouldn't be required given that the google colab version is running on cards with v12.* of CUDA. Unless there is a major difference  between what is running on those notebooks and what's publicly released, this shouldn't be the issue.
